src/mainwin.cc

```diff
@@ -1027,83 +1027,89 @@ void MainWindow::refreshDocument()
+					shouldcompiletoplevel = true;
+				}
 			}
 		}
-		// If the file hasn't changed, we might still need to compile it
-		// if we haven't yet compiled the current text.
 		else {
-			auto current_doc = editor->toPlainText();
-			if (current_doc != last_compiled_doc && last_compiled_doc.size() == 0) {
+			shouldcompiletoplevel = true;
+		}
+
+		if (!shouldcompiletoplevel && this->parsed_module) {
+			auto mtime = this->parsed_module->includesChanged();
+			if (mtime > this->includes_mtime) {
+				this->includes_mtime = mtime;
 				shouldcompiletoplevel = true;
 			}
 		}
-	}
-	else {
-		shouldcompiletoplevel = true;
-	}
 
-	if (!shouldcompiletoplevel && this->parsed_module) {
-		auto mtime = this->parsed_module->includesChanged();
-		if (mtime > this->includes_mtime) {
-			this->includes_mtime = mtime;
-			shouldcompiletoplevel = true;
+		if (shouldcompiletoplevel) {
+			if (editor->isContentModified()) saveBackup();
+			compileTopLevelDocument(rebuildParameterWidget);
+			didcompile = true;
 		}
-	}
 
-	if (shouldcompiletoplevel) {
-		if (editor->isContentModified()) saveBackup();
-		compileTopLevelDocument(rebuildParameterWidget);
-		didcompile = true;
-	}
-
-	if (this->root_module) {
-		auto mtime = this->root_module->handleDependencies();
-		if (mtime > this->deps_mtime) {
-			this->deps_mtime = mtime;
-			PRINTB("Used file cache size: %d files", ModuleCache::instance()->size());
-			didcompile = true;
+		if (this->root_module) {
+			auto mtime = this->root_module->handleDependencies();
+			if (mtime > this->deps_mtime) {
+				this->deps_mtime = mtime;
+				PRINTB("Used file cache size: %d files", ModuleCache::instance()->size());
+				didcompile = true;
+			}
 		}
-	}
 
-	// If we're auto-reloading, listen for a cascade of changes by starting a timer
-	// if something changed _and_ there are any external dependencies
-	if (reload && didcompile && this->root_module) {
-		if (this->root_module->hasIncludes() ||
-				this->root_module->usesLibraries()) {
-			this->waitAfterReloadTimer->start();
-			this->procevents = false;
-			return;
+		// If we're auto-reloading, listen for a cascade of changes by starting a timer
+		// if something changed _and_ there are any external dependencies
+		if (reload && didcompile && this->root_module) {
+			if (this->root_module->hasIncludes() ||
+					this->root_module->usesLibraries()) {
+				this->waitAfterReloadTimer->start();
+				this->procevents = false;
+				return;
+			}
 		}
-	}
 
-	if (!reload && didcompile) {
-		if (!animate_panel->isVisible()) {
-			emit unhighlightLastError();
-			if (!this->root_module) {
-				emit highlightError( parser_error_pos );
+		if (!reload && didcompile) {
+			if (!animate_panel->isVisible()) {
+				emit unhighlightLastError();
+				if (!this->root_module) {
+					emit highlightError( parser_error_pos );
+				}
 			}
 		}
-	}
 
-	compileDone(didcompile | forcedone);
+		compileDone(didcompile | forcedone);
+	}catch(HardWarningException){
```

Comment:
`const HardWarningException&`

..because we don't want to make a copy of an exception. and we may not want to even know if it's a concrete class.